src/util/hex.1
.TH HEX 1 BRL-CAD
.\"                           H E X . 1
.\" BRL-CAD
.\"
.\" Copyright (c) 1989-2012 United States Government as represented by
.\" the U.S. Army Research Laboratory.
.\"
.\" Redistribution and use in source (Docbook format) and 'compiled'
.\" forms (PDF, PostScript, HTML, RTF, etc.), with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\"
.\" 1. Redistributions of source code (Docbook format) must retain the
.\" above copyright notice, this list of conditions and the following
.\" disclaimer.
.\"
.\" 2. Redistributions in compiled form (transformed to other DTDs,
.\" converted to PDF, PostScript, HTML, RTF, and other formats) must
.\" reproduce the above copyright notice, this list of conditions and
.\" the following disclaimer in the documentation and/or other
.\" materials provided with the distribution.
.\"
.\" 3. The name of the author may not be used to endorse or promote
.\" products derived from this documentation without specific prior
.\" written permission.
.\"
.\" THIS DOCUMENTATION IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY
.\" EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
.\" PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT
.\" OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
.\" BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
.\" (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
.\" USE OF THIS DOCUMENTATION, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\".\".\"
.SH NAME
hex \- Hexadecimal and ASCII dump
.SH SYNOPSIS
.B hex
.I
[
.I \-o offset
]
[
.I file .\|.\|.
]
.SH DESCRIPTION
.I Hex
reads each
.I file
in sequence and
produces a dump on the standard output
in hexadecimal and ASCII.  If no files are specified on the command
line,
.I hex
checks to to be sure the standard input is not coming from a terminal
before dumping from the standard input.
.LP
Input is
dumped 16 bytes at a time.  The program displays the offset
from the beginning of the file/input
in hexadecimal, followed by
the hexadecimal representations for the
next 16
bytes of data, and their ASCII character representations.  Non-printing
characters are displayed as a period (".") in the ASCII section.  Vertical
bars delimit the ASCII section.
.LP
If multiple files are specified on the command line, each file's name is
printed before its dump.
.LP
.SH OPTIONS
.TP
.B \-o offset
start printing at byte position
.I offset
instead of at the beginning of the data.
.I Offset
is rounded down to a multiple of 16.
.SH SEE ALSO
od
.SH AUTHOR
BRL-CAD Team
.SH COPYRIGHT
This software is Copyright (c) 1989-2012 by the United States
Government as represented by U.S. Army Research Laboratory.
.SH "BUG REPORTS"
Reports of bugs or problems should be submitted via electronic
mail to <devs@brlcad.org>.


NEWS
						 -*- coding: utf-8 -*-
BRL-CAD Release Notes
=====================

This document provides a transcript of release notes and user-visible
changes for each release of BRL-CAD.  The individual changes are
listed in chronological order with the most recent changes listed
first and are grouped by release.  Each release also includes optional
descriptive text that emphasizes or further describes significant
changes made.  See document footer for additional details.

----------------------------------------------------------------------
--- 2012-07-XX  Release 7.22.2                                     ---
----------------------------------------------------------------------

* fixed corrupted tessellation pipe export bug - Richard Weiss
* output from g-var no longer prints 'fwrite' messages - Sean Morrison
* fixed crash in step-g tool importing AP214 entities - Sean Morrison
* improved 3dm-g output and invalid geometry support - Sean Morrison
* fixed crash revolving empty sketch objects - Sean Morrison
* fixed TIE raytracing bugs shotlining plate-mode BoTs - Sean Morrison
* g_diff gives non-zero return code if .g files differ - Sean Morrison
* fixed typographical errors in numerous tools and docs - Carl Moore
* fixed TIE perpendicular face mesh evaluation failure - Richard Weiss


----------------------------------------------------------------------
--- 2012-06-25  Release 7.22.0                                     ---
----------------------------------------------------------------------

After many months of development, this release includes hundreds of
fixes for source code defects, logic and memory management bugs,
potential security vulnerabilities, and numerous other issues reported
by Coverity Static Analysis.  While having participated in the
Coverity Scan Initiative since inception in 2006, this release marks a
completion milestone with the review and resolution of all reported
issues, more than 2100 overall.  Every defect was inspected, code
modifications were peer-reviewed, critical features were tested, and
user-visible changes were documented.  Along with other on-going
improvement efforts such as treating all compiler-generated warnings
as errors, time and attention continues to be put forth towards
improving BRL-CAD's maintainability and source code quality.

Development of BREP/NURBS representation support within BRL-CAD
continues to mature with this release including ray tracing
improvements, tighter bounding boxes for better performance, progress
converting from traditional implicit+CSG representations to NURBS+CSG,
wireframe improvements, and a thorough verification and validation
(V&V) review for analysis suitability.  The V&V review compared
BRL-CAD's primitives in implicit, polygonal mesh, and NURBS form
evaluating more than a dozen separate tests looking for differences in
rendering, curvature, presented areas, volume, and more.  The review
identified a couple minor issues that were inspected and fixed but
successfully demonstrated the viability of using our BREP/NURBS
implementation for geometric representation and analysis purposes.

This release introduces numerous major feature developments including
the initial implementation of a new physics simulation capability, a
new ambient occlusion rendering mode, and the integration of rtwizard
into Archer.  The simulation system was developed by Abhijit Nandy
under the European Space Agency's 2011 Summer of Code in Space (SOCIS)
program; and it allows BRL-CAD geometry objects to dynamically
interact with each other (e.g., collision detection, friction) or with
the environment (e.g., gravity).  Ambient occlusion is a shading
method that helps add considerable realism and feature definition to
renderings, enabled within rt using the "set ambSamples=128" command.
The rtwizard rendering interface was extended to work as a
command-line tool (useful for scripting) and was integrated within
Archer as the underlying rendering interface.

Conversion support continues to be a development priority with
significant importer enhancements for the STEP, OBJ, COMGEOM, FASTGEN,
and RAW file formats.  The STEP importer in particular receives
several updates, support for importing assembly hierarchies, and is in
the process of being fully ported to Windows.  This release also fixes
more than 40 user-visible bugs including a significant database
corruption issue.  A database corruption issue was discovered and
fixed affecting 64-bit big endian platforms (e.g., IBM Power Series
supercomputers) when writing out ebm, dsp, vol, and hf primitives.

BRL-CAD's documentation continues to be extended, improved, and
translated due to the sustained efforts of several contributors.
Manual page documentation is now available on Windows through the
'brlman' command and within MGED.  Several utilities and commands that
previously had no documentation now have manual pages; many others
received corrections and improvements.  The introductory article
"About BRL-CAD" is now available in Russian, Italian, and Armenian.

* numerous manual page corrections and improvements - Tom Browder
* integration of rtwizard traits into archer - Bob Parker, Cliff Yapp
* new asc2dsp prep tool for displacement map data files - Tom Browder
* fixed a problem with rt writing pix files on Windows - Bob Parker
* added new raw-g raw triangle mesh to BoT importer - Daniel Roßberg
* added Windows man page support (using html) to brlman - Cliff Yapp
* added support for section numbers to brlman script - Cliff Yapp
* new manual pages for asc2dsp, dsp, asc2pix, and asc2g - Tom Browder
* fixed bug creating binary uniform (binunif) objects - Sean Morrison
* fixed memory corruption bug reading v4 geometry - Sean Morrison
* fixed bug where small raytraces were being rotated - Erik Greenwald
* ios-icon.sh script added for generating IOS icons - Erik Greenwald
* support for implicit to boundary representation geometry conversion
	- Cliff Yapp, Wu Jianbang, Sean Morrison

src/util/hex.c
/*                            H E X . C
 * BRL-CAD
 *
 * Copyright (c) 2004-2012 United States Government as represented by
 * the U.S. Army Research Laboratory.
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * version 2.1 as published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this file; see the file named COPYING for more
 * information.
 */
/** @file util/hex.c
 *
 * Give a good ole' CPM style Hex dump of a file or standard input.
 *
 * Options
 * h    help
 * o    offset from beginning of data from which to start dump
 */

#include "common.h"

#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include "bio.h"

#include "bu.h"


/* declarations to support use of bu_getopt() system call */
static char *options = "o:";
static char *progname = "(noname)";

static long offset=0;	 /* offset from beginning of file from which to start */

#define DUMPLEN 16    /* number of bytes to dump on one line */

/*
 * D U M P --- Dump file in hex
 */
void dump(FILE *fd)
{
    int i;
    char *p;
    int bytes;
    long addr = 0L;
    static char buf[DUMPLEN];    /* input buffer */

    if (offset != 0) {
	/* skip over "offset" bytes first */
	if (fseek(fd, offset, 0)) {

	    /* If fseek fails, try reading our way to the desired offset.
	     * The fseek will fail if we're reading from a pipe.
	     */

	    addr=0;
	    while (addr < offset) {
		if ((i=fread(buf, 1, sizeof(buf), fd)) == 0) {
		    fprintf(stderr, "%s: offset exceeds end of input!\n", progname);
		    bu_exit (-1, NULL);
		} else addr += i;
	    }
	} else addr = offset;
    }

    /* dump address, Hex of buffer and ASCII of buffer */
    while ((bytes=fread(buf, 1, sizeof(buf), fd)) > 0) {

	/* print the offset into the file */
	printf("%08lx", addr);

	/* produce the hexadecimal dump */
	for (i=0, p=buf; i < DUMPLEN; ++i) {
	    if (i < bytes) {
		if (i%4 == 0) printf("  %02x", *p++ & 0x0ff);
		else printf(" %02x", *p++ & 0x0ff);
	    } else {
		if (i%4 == 0) printf("    ");
		else printf("   ");
	    }
	}

	/* produce the ASCII dump */
	printf(" |");
	for (i=0, p=buf; i < bytes; ++i, ++p) {
	    int c = *p;
	    if (c < 0)
		c = 0;
	    if (c > 255)
		c = 255;
	    if (isascii(c) && isprint(c))
		putchar(c);
	    else
		putchar('.');
	}
	printf("|\n");
	addr += DUMPLEN;
    }
}


/*
 * U S A G E --- Print helpful message and bail out
 */
void usage(void)
{
    (void) fprintf(stderr, "Usage: %s [-o offset] [file...]\n", progname);
    bu_exit (1, NULL);
}


/* M A I N
 *
 * Parse arguments and call 'dump' to perform primary task.
 */
int
main(int ac, char **av)
{
    int c, optlen, files;
    FILE *fd;
    char *eos;
    long newoffset;

    progname = *av;

    /* Get # of options & turn all the option flags off */
    optlen = strlen(options);

    for (c=0; c < optlen; c++)  /* NIL */;

    /* Turn off bu_getopt's error messages */
    bu_opterr = 0;

    /* get all the option flags from the command line */
    while ((c=bu_getopt(ac, av, options)) != -1)
	if (c == 'o') {
	    newoffset = strtol(bu_optarg, &eos, 0);

	    if (eos != bu_optarg)
		offset = newoffset;
	    else
		fprintf(stderr, "%s: error parsing offset \"%s\"\n",
			progname, bu_optarg);
	} else usage();

    if (offset%DUMPLEN != 0) offset -= offset % DUMPLEN;

    if (bu_optind >= ac) {
	/* no file left, try processing stdin */
	if (isatty(fileno(stdin))) usage();
	else dump(stdin);
    } else {
	/* process each remaining arguments */
	for (files = ac-bu_optind; bu_optind < ac; bu_optind++) {
	    if ((fd=fopen(av[bu_optind], "r")) == (FILE *)NULL) {
		perror(av[bu_optind]);
		bu_exit (-1, NULL);
	    }
	    if (files > 1) printf("/**** %s ****/\n", av[bu_optind]);
	    dump(fd);
	    (void)fclose(fd);
	}
    }
    return 0;
}


/*
 * Local Variables:
 * mode: C
 * tab-width: 8
 * indent-tabs-mode: t
 * c-file-style: "stroustrup"
 * End:
 * ex: shiftwidth=4 tabstop=8
 */
